File: go/test/sandbox_filesystem_snapshot_test.go
package test

import (
	"io"
	"testing"
	"time"

	modal "github.com/modal-labs/modal-client/go"
	"github.com/onsi/gomega"
)

func TestSnapshotFilesystem(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)
	ctx := t.Context()
	tc := newTestClient(t)

	app, err := tc.Apps.FromName(ctx, "libmodal-test", &modal.AppFromNameParams{CreateIfMissing: true})
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	image := tc.Images.FromRegistry("alpine:3.21", nil)

	sb, err := tc.Sandboxes.Create(ctx, app, image, nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())
	sbFromID, err := tc.Sandboxes.FromID(ctx, sb.SandboxID)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())
	defer terminateSandbox(g, sbFromID)

	writeFile, err := sb.Exec(ctx, []string{"sh", "-c", "echo -n 'test content' > /tmp/test.txt"}, nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())
	_, err = writeFile.Wait(ctx)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	mkDir, err := sb.Exec(ctx, []string{"mkdir", "-p", "/tmp/testdir"}, nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())
	_, err = mkDir.Wait(ctx)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	snapshotImage, err := sb.SnapshotFilesystem(ctx, 55*time.Second)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())
	g.Expect(snapshotImage).ShouldNot(gomega.BeNil())
	g.Expect(snapshotImage.ImageID).To(gomega.HavePrefix("im-"))

	_, err = sb.Terminate(ctx, nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	sb2, err := tc.Sandboxes.Create(ctx, app, snapshotImage, nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())
	defer terminateSandbox(g, sb2)

	proc, err := sb2.Exec(ctx, []string{"cat", "/tmp/test.txt"}, nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	output, err := io.ReadAll(proc.Stdout)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())
	g.Expect(string(output)).To(gomega.Equal("test content"))

	dirCheck, err := sb2.Exec(ctx, []string{"test", "-d", "/tmp/testdir"}, nil)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())

	exitCode, err := dirCheck.Wait(ctx)
	g.Expect(err).ShouldNot(gomega.HaveOccurred())
	g.Expect(exitCode).To(gomega.Equal(0))
}

File: js/test/sandbox_filesystem_snapshot.test.ts
import { tc } from "../test-support/test-client";
import { expect, test, onTestFinished } from "vitest";

test("snapshotFilesystem", async () => {
  const app = await tc.apps.fromName("libmodal-test", {
    createIfMissing: true,
  });
  const image = tc.images.fromRegistry("alpine:3.21");

  const sb = await tc.sandboxes.create(app, image);

  const writeFile = await sb.exec([
    "sh",
    "-c",
    "echo -n 'test content' > /tmp/test.txt",
  ]);
  await writeFile.wait();
  const mkDir = await sb.exec(["mkdir", "-p", "/tmp/testdir"]);
  await mkDir.wait();

  const snapshotImage = await sb.snapshotFilesystem();
  expect(snapshotImage.imageId).toMatch(/^im-/);

  await sb.terminate();

  // Create new Sandbox from snapshot
  const sb2 = await tc.sandboxes.create(app, snapshotImage);
  onTestFinished(async () => await sb2.terminate());

  const proc = await sb2.exec(["cat", "/tmp/test.txt"]);
  const output = await proc.stdout.readText();
  expect(output).toBe("test content");

  const dirCheck = await sb2.exec(["test", "-d", "/tmp/testdir"]);
  expect(await dirCheck.wait()).toBe(0);
});
